### PR TITLE
[Core][Dashboard Pubsub 1/n] Allow a channel to have subscribers to a key and to the whole channel concurrently

### PR DESCRIPTION
**Pull request by mwtian**
## Why are these changes needed?
For actor channel, GCS clients subscribe to a single actor but dashboard subscribes to all actors. This change makes supporting this possible.

Most of the added code is in `integration_test.cc`, which tests the publisher and subscriber together.

Also, add the basic support for dashboard reporter pubsub.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/pubsub/subscriber.cc [hunk 1]
```diff
@@ -219,6 +219,15 @@ std::string SubscriberChannel::DebugString() const {
 /// Subscriber
 ///////////////////////////////////////////////////////////////////////////////
 
+Subscriber::~Subscriber() {
+  if (!mutex_.LockWhenWithTimeout(absl::Condition(this, &Subscriber::CheckNoLeaks_Locked),
```
Comment (rkooo567) [inline src/ray/pubsub/subscriber.cc:223]: Should we do this only when debug mode is on?

File: src/ray/gcs/gcs_client/gcs_client.cc [hunk 2]
```diff
@@ -145,7 +145,7 @@ Status GcsClient::Connect(instrumented_io_context &io_service) {
   rpc::Address gcs_address;
   gcs_address.set_ip_address(current_gcs_server_address_.first);
   gcs_address.set_port(current_gcs_server_address_.second);
-  /// TODO: refactor pubsub::Subscriber to avoid this.
+  /// TODO(mwtian): refactor pubsub::Subscriber to avoid this.
```
[the remark this answers is not in the document]
Comment (mwtian) [reply src/ray/gcs/gcs_client/gcs_client.cc:148]: Reworded. Basically faking a worker ID for GCS seems confusing.

File: src/ray/protobuf/reporter.proto [hunk 3]
```diff
@@ -52,6 +52,14 @@ message ReportOCMetricsRequest {
 message ReportOCMetricsReply {
 }
 
+// Resource usage reported by the node reporter.
+message NodeResourceUsage {
```
[the remark this answers is not in the document]
Comment (mwtian) [reply src/ray/protobuf/reporter.proto:56]: This proto `NodeResourceUsage` seems more relevant to reporters. Also `common.proto` is relative long now with a mix of proto definitions, making dependency management harder.

File: src/ray/pubsub/publisher.cc [hunk 4]
```diff
@@ -23,53 +23,47 @@ namespace pub_internal {
 bool SubscriptionIndex::AddEntry(const std::string &key_id,
                                  const SubscriberID &subscriber_id) {
   if (key_id.empty()) {
-    // This subscriber will subscribe to all entities in the channel.
```
[the remark this answers is not in the document]
Comment (mwtian) [reply src/ray/pubsub/publisher.cc:26]: For actor channel, there are both subscribers to a single actor ID (from GCS client), and subscribers to all actor IDs from dashboard. We need to support this case.

File: src/ray/pubsub/publisher.cc [hunk 5]
```diff
@@ -23,53 +23,47 @@ namespace pub_internal {
...
+    auto &ids = it->second;
+    subscribers.insert(subscribers.end(), ids.begin(), ids.end());
   }
-  return absl::optional<std::reference_wrapper<const absl::flat_hash_set<SubscriberID>>>{
-      std::ref(it->second)};
+  return subscribers;
 }
 
 bool SubscriptionIndex::EraseSubscriber(const SubscriberID &subscriber_id) {
-  if (!subscribers_to_all_.empty()) {
```
[the remark this answers is not in the document]
Comment (mwtian) [reply src/ray/pubsub/publisher.cc:62]: Yeah, a lot of refactoring, and supporting erasing a subscriber, when the channel has both single-ID and all-ID subscribers.

File: src/ray/pubsub/subscriber.cc (hunk 1 above)
Comment (mwtian) [reply src/ray/pubsub/subscriber.cc:223]: Removing this for now. In future it will be a debug only check.

File: src/ray/pubsub/publisher.cc [hunk 6]
```diff
@@ -23,53 +23,47 @@ namespace pub_internal {
...
 
   RAY_CHECK(key_added == subscriber_added);
   return key_added;
 }
 
-absl::optional<std::reference_wrapper<const absl::flat_hash_set<SubscriberID>>>
-SubscriptionIndex::GetSubscriberIdsByKeyId(const std::string &key_id) const {
+std::vector<SubscriberID> SubscriptionIndex::GetSubscriberIdsByKeyId(
+    const std::string &key_id) const {
+  std::vector<SubscriberID> subscribers;
```
Comment (iycheng) [inline src/ray/pubsub/publisher.cc:40]: could it lead t some perf issue?

File: src/ray/pubsub/publisher.cc [hunk 7]
```diff
@@ -81,20 +75,18 @@ bool SubscriptionIndex::EraseSubscriber(const SubscriberID &subscriber_id) {
       key_id_to_subscribers_.erase(subscribers_it);
     }
   }
-  subscribers_to_key_id_.erase(subscribing_message_it);
+  subscribers_to_key_id_.erase(subscribing_key_it);
   return true;
 }
 
 bool SubscriptionIndex::EraseEntry(const std::string &key_id,
```
Comment (iycheng) [inline src/ray/pubsub/publisher.cc:82]: Not related to this PR, but I feel the doc string should be improved [here](https://github.com/ray-project/ray/blob/41f80f641cf7e3587fadaf6ea3e79c5258df760f/src/ray/pubsub/publisher.h#L58-L59)

For example, if key_id == "", what doat that mean?

**Review by fishbone — COMMENTED** [review]
Overall I'm good about this. Some comments.

File: src/ray/pubsub/publisher.cc (hunk 6 above)
Comment (mwtian) [reply src/ray/pubsub/publisher.cc:40]: I think this is unlikely to be hot, relative to other cost e.g. copying the PubMessage proto.

File: src/ray/pubsub/publisher.cc [hunk 8]
```diff
@@ -296,6 +288,17 @@ bool Publisher::UnregisterSubscriber(const SubscriberID &subscriber_id) {
...
 
+void Publisher::UnregisterAll() {
+  absl::MutexLock lock(&mutex_);
+  std::vector<SubscriberID> ids;
+  for (const auto &[id, subscriber] : subscribers_) {
+    ids.push_back(id);
+  }
+  for (const auto &id : ids) {
+    UnregisterSubscriberInternal(id);
+  }
```
[the remark this answers is not in the document]
Comment (mwtian) [reply src/ray/pubsub/publisher.cc:299]: Added a comment. Basically `UnregisterSubscriberInternal(id)` erases from the `subscribers_` hashmap so we have to save the keys first.

File: src/ray/pubsub/publisher.cc (hunk 7 above)
Comment (mwtian) [reply src/ray/pubsub/publisher.cc:82]: Added comment.

**Review by fishbone — APPROVED** [review]
LGTM. Waiting @rkooo567 for another review.

**Review by rkooo567 — APPROVED** [review]

**Pull request merged**